nifi-nar-bundles/nifi-extension-utils/nifi-processor-utils/src/main/java/org/apache/nifi/processor/util/list/ListedEntityTracker.java

```diff
@@ -0,0 +1,322 @@
+                return false;
+            }
+
+            final ListedEntity alreadyListedEntity = alreadyListedEntities.get(identifier);
+            if (alreadyListedEntity == null) {
+                logger.trace("Picked {} being newly found.", new Object[]{identifier});
+                return true;
+            }
+
+            if (entity.getTimestamp() > alreadyListedEntity.getTimestamp()) {
+                logger.trace("Picked {} having newer timestamp {} than {}.",
+                        new Object[]{identifier, entity.getTimestamp(), alreadyListedEntity.getTimestamp()});
+                return true;
+            }
+
+            if (entity.getSize() != alreadyListedEntity.getSize()) {
+                logger.trace("Picked {} having different size {} than {}.",
+                        new Object[]{identifier, entity.getSize(), alreadyListedEntity.getSize()});
+                return true;
+            }
+
+            logger.trace("Skipped {}, not changed.", new Object[]{identifier, entity.getTimestamp(), minTimestampToList});
+            return false;
+        }).collect(Collectors.toList());
+
+        // Find old enough entries.
+        final List<String> oldEntityIds = alreadyListedEntities.entrySet().stream()
+                .filter(entry -> entry.getValue().getTimestamp() < minTimestampToList).map(Map.Entry::getKey)
+                .collect(Collectors.toList());
+
+        if (updatedEntities.isEmpty() && oldEntityIds.isEmpty()) {
+            logger.debug("None of updated or old entity was found. Yielding.");
+            context.yield();
+            return;
+        }
+
+        // Remove old entries.
+        oldEntityIds.forEach(oldEntityId -> alreadyListedEntities.remove(oldEntityId));
+
+        // Emit updated entities.
+        updatedEntities.forEach(updatedEntity -> {
+            FlowFile flowFile = session.create();
+            flowFile = session.putAllAttributes(flowFile, createAttributes.apply(updatedEntity));
+            session.transfer(flowFile, REL_SUCCESS);
+            // In order to reduce object size, discard meta data captured at the sub-classes.
+            final ListedEntity listedEntity = new ListedEntity();
+            listedEntity.setTimestamp(updatedEntity.getTimestamp());
+            listedEntity.setSize(updatedEntity.getSize());
+            alreadyListedEntities.put(updatedEntity.getIdentifier(), listedEntity);
+        });
+
+        // Commit ProcessSession before persisting listed entities.
+        // In case persisting listed entities failure, same entities may be listed again, but better than not listing.
+        session.commit();
+        try {
+            logger.debug("Removed old entities count: {}, Updated entities count: {}",
+                    new Object[]{oldEntityIds.size(), updatedEntities.size()});
+            if (logger.isTraceEnabled()) {
+                logger.trace("\"Removed old entities: {}, Updated entities: {}",
```

Comment:
Was this log message intended to begin with a double-quote?

Reply:
Thanks, it was copy-pasting. Removed the double-quote.